### PR TITLE
Initial version of test status bot

### PR DESCRIPTION
**Pull request by danielrozenberg**
Part of the implementation of https://github.com/ampproject/amphtml/issues/20744

This is a pretty large PR, so I split it into more consumable chunks in separate commits. Take your time reviewing!

### CODE REVIEW

File: test-status/README.md [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+* `POST /v0/tests/:headSha/:type/:status(queued|started|skipped)`
+  * Creates a new check on the supplied head commit (for `status` = `queued` or
+    `skipped`) or reports that the `type` tests have started running.
+* `POST /v0/tests/:headSha/:type/report/:passed/:failed`
+  * Updates the equivalent check with the number of `passed` and `failed` tests
+  * If `failed` is 0, sets the check's conclusion to `success`, which turns the
+    check green
+  * If `failed` ≥ 1, sets the check's conclusion to `action_required`, which
+    turns the check red. It also sets a URL to resolve the issue back to the web
+    interface of this app, which can only be accessed by the weekly build cop
```
Comment (rsimha) [inline test-status/README.md]: I fear that allowing only the build-cop to resolve a failed run will form a bottleneck. There are likely to be genuine cases where the PR author / reviewer / owner would want to approve a test run. Perhaps we can start with a well-known set of approvers (start with everyone on the infra team) for flaky test runs, and also include the build cop in that list?

File: test-status/api.js [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+    head_sha: pullRequestSnapshot.head_sha,
+  };
+  switch (status) {
+    case 'queued':
+      Object.assign(params, {
+        status: 'queued',
+        output: {
+          title: `${type} tests are queued on Travis`,
+          summary: `The ${type} tests are queued to run on Travis. Watch ` +
+            'this space for results in a few minutes!',
```
Comment (rsimha) [inline test-status/api.js]: I think most of this message will be truncated. Perhaps shorten to something like: `${type} tests queued on Travis. Waiting for status...`?

image

File: test-status/api.js [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+      break;
+
+    case 'started':
+      Object.assign(params, {
+        status: 'in_progress',
+        started_at: new Date().toISOString(),
+        output: {
+          title: `${type} tests are running on Travis`,
+          summary: `The ${type} tests are running on Travis. Watch this ` +
+            'space for results in a few minutes!',
```
Comment (rsimha) [inline test-status/api.js]: Ditto.

File: test-status/api.js [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+        `/tests/${headSha}/${type}/status`, process.env.WEB_UI_BASE_URL);
+    Object.assign(params, {
+      details_url: detailsUrl.href,
+      conclusion: 'action_required',
+      output: {
+        title: `${failed} ${type} test${failed != 1 ? 's' : ''} failed`,
+        summary: `The ${type} tests finished running on Travis.`,
+        text: `* *${passed}* test${passed != 1 ? 's' : ''} PASSED\n` +
+          `* *${failed}* test${failed != 1 ? 's' : ''} FAILED\n\n` +
+          'Please inspect the Travis build and fix any code changes that ' +
```
Comment (rsimha) [inline test-status/api.js]: Any chance you can link to the Travis job here?

File: test-status/webhooks.js [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+'use strict';
+
+exports.installGitHubWebhooks = (app, db) => {
+  app.on(['pull_request.opened', 'pull_request.synchronize'], async context => {
```
Comment (rsimha) [inline test-status/webhooks.js]: Will the status get updated when an existing Travis job is re-run? (Sorry, I'm still getting familiarized with the flow here.)

File: test-status/README.md (hunk 1 above)
Comment (danielrozenberg) [reply test-status/README.md]: It's not in the README, but yet - there is also a list of fixed approvers (currently I set it to the infra team, later we can pass the torch to others.) As I mentioned offline, the logic to determine who the build cop is is not there yet :laughing: 

I'll add a mentioned of the fixed list to the README.md though

File: test-status/api.js (hunk 2 above)
Comment (danielrozenberg) [reply test-status/api.js]: The title is the short text that appears in the screenshot about - the summary is what's hidden behind the "details" link. Even if it gets truncated, I'm confident the first 4 works (`${type} tests are queued…`) will be visible

File: test-status/api.js (hunk 4 above)
Comment (danielrozenberg) [reply test-status/api.js]: #85

File: test-status/webhooks.js (hunk 5 above)
Comment (danielrozenberg) [reply test-status/webhooks.js]: Yes

**Pull request merged**